src/main.cpp

```diff
@@ -2398,11 +2398,16 @@ bool static DisconnectTip(CValidationState &state) {
     mempool.check(pcoinsTip);
     // Update chainActive and related variables.
     UpdateTip(pindexDelete->pprev);
+    // Get the current commitment tree
+    ZCIncrementalMerkleTree newTree;
+    assert(pcoinsTip->GetAnchorAt(pcoinsTip->GetBestAnchor(), newTree));
     // Let wallets know transactions went from 1-confirmed to
     // 0-confirmed or conflicted:
     BOOST_FOREACH(const CTransaction &tx, block.vtx) {
         SyncWithWallets(tx, NULL);
     }
+    // Update cached incremental witnesses
+    GetMainSignals().ChainTip(pindexDelete, &block, newTree, false);
```

Comment:
Locking of what? We lock the wallet inside `IncrementNoteWitnesses()` and `DecrementNoteWitnesses()` which are called inside `ChainTip()`. AFAICT my usage of `GetMainSignals()` is consistent with other uses in `main.cpp`.